src/session.c

```diff
@@ -7,33 +7,168 @@
+{
+    P11PROV_SESSION_POOL *pool;
+
+    pool = OPENSSL_zalloc(sizeof(P11PROV_SESSION_POOL));
+    if (!pool) {
+        return CKR_HOST_MEMORY;
+    }
+    pool->provctx = ctx;
+
+    pthread_mutex_init(&pool->lock, 0);
+
+    if (token->ulMaxSessionCount != 0) {
+        pool->max_sessions = token->ulMaxSessionCount;
+        /* keep a max of 10% of the sessions */
+        pool->limit_sessions = pool->max_sessions / 10;
+        if (pool->limit_sessions == 0) {
+            pool->limit_sessions = 3;
+        }
+    } else {
+        /* arbitrary max concurrent open sessions */
+        pool->max_sessions = 1024;
+        pool->limit_sessions = 8;
+    }
+    if (pool->limit_sessions > pool->max_sessions) {
+        pool->limit_sessions = pool->max_sessions;
+    }
+
+    *_pool = pool;
+    return CKR_OK;
+}
+
+CK_RV p11prov_session_pool_free(P11PROV_SESSION_POOL *pool)
+{
+    P11PROV_SESSION *session;
+
+    if (!pool) {
+        return CKR_OK;
+    }
+
+    /* deref login_session first */
+    p11prov_session_free(pool->login_session);
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&pool->lock);
+    for (int i = 0; i < pool->num_sessions; i++) {
+        session = pool->sessions[i];
+        if (session->refcnt > 1) {
+            P11PROV_raise(pool->provctx, CKR_GENERAL_ERROR,
+                          "Can't free multiply-referenced session (%u)", i);
+        }
+        if (session->session != CK_INVALID_HANDLE) {
+            CK_FUNCTION_LIST *f;
+            int ret;
+
+            f = p11prov_ctx_fns(pool->provctx);
+            if (f != NULL) {
+                ret = f->C_CloseSession(session->session);
+                if (ret != CKR_OK) {
+                    P11PROV_raise(pool->provctx, ret,
+                                  "Failed to close session %lu",
+                                  session->session);
+                }
+            }
+            (void)__atomic_fetch_sub(&pool->cur_sessions, 1, __ATOMIC_SEQ_CST);
+        }
+        OPENSSL_clear_free(session, sizeof(P11PROV_SESSION));
+        pool->sessions[i] = NULL;
+    }
+    OPENSSL_free(pool->sessions);
+    pool->sessions = NULL;
+    pool->num_sessions = 0;
+    pthread_mutex_unlock(&pool->lock);
+    /* ------------- LOCKED SECTION */
+
+    pthread_mutex_destroy(&pool->lock);
+    OPENSSL_clear_free(pool, sizeof(P11PROV_SESSION_POOL));
+
+    return CKR_OK;
+}
+
+#define SESS_ALLOC_SIZE 8
+static P11PROV_SESSION *p11prov_session_new(P11PROV_CTX *ctx,
+                                            struct p11prov_slot *slot)
 {
-    P11PROV_SESSION *sess;
+    P11PROV_SESSION_POOL *p = slot->pool;
+    P11PROV_SESSION *session;
+    CK_ULONG cur_sessions;
+
+    /* this is just to quit early and avoid work.
+     * Passing this check does not guarantee another thread is not
+     * racing to open the last avaliable session */
+    cur_sessions = __atomic_load_n(&p->cur_sessions, __ATOMIC_SEQ_CST);
```

Comment:
Should this be `num_sessions` instead? This checks against pkcs11 session count which can get adjusted in various places. `num_sessions` is more monotonic on how many internal session objects we have.

Reply:
No because internal object may not have sessions associated, we really want to check against the actually open sessions with the token.